python/lsst/pipe/analysis/utils.py

```diff
@@ -888,10 +889,81 @@ def calibrateSourceCatalogMosaic(dataRef, catalog, fluxKeys=None, errKeys=None,
+    """Calibrate catalog with (e.g. jointcal/meas_mosaic) PhotoCalib results
+
+    Parameters
+    ----------
+    dataRef : `lsst.daf.persistence.butlerSubset.ButlerDataRef`
+       The data reference for which the relevant datasets are to be retrieved.
+    catalog : `lsst.afw.table.SourceCatalog`
+       The source catalog to which the calibrations will be applied.
+    fluxKeys : `dict`, optional
+       A `dict` of the flux keys to which the photometric calibration will
+       be applied.  If not provided, the getFluxKeys function will be used
+       to set it.
+    zp : `float`, optional
+       A constant zero point magnitude to which to scale all the fluxes in
+       ``fluxKeys``.
+
+    Returns
+    -------
+    catalog : `lsst.afw.table.SourceCatalog`
+       The calibrated catalog.
+    """
+    wcs = dataRef.get("jointcal_wcs")
+    for record in catalog:
+        record.updateCoord(wcs)
+
+    photoCalib = dataRef.get("jointcal_photoCalib")
+    # TODO: need to scale these until DM-10153 is completed and PhotoCalib has
+    # replaced Calib entirely
+    referenceFlux = 1e23*10**(48.6/-2.5)*1e9
+    # Convert to constant zero point, as for the coadds
+    factor = referenceFlux/10.0**(0.4*zp)
+    if fluxKeys is None:
+        fluxKeys, errKeys = getFluxKeys(catalog.schema)
+
+    for fluxName, fluxKey in list(fluxKeys.items()):
+        if len(catalog[fluxKey].shape) > 1:
+            continue
+        try:  # photoCalib.instFluxToNanojansky() requires an error for each flux
+            fluxErrKey = catalog.schema.find(fluxName + "Err").key
+        except KeyError:
+            fluxErrKey = None
+        baseName = re.sub("_instFlux", "", fluxName)
+        if fluxErrKey:
+            calibratedFluxAndErrArray = photoCalib.instFluxToNanojansky(catalog, baseName)
+            catalog[fluxKey] = calibratedFluxAndErrArray[:, 0]
+            catalog[fluxErrKey] = calibratedFluxAndErrArray[:, 1]
+        else:
+            # photoCalib requires an error for each flux, but some don't
+            # have one in the schema (currently only certain deblender
+            # fields, e.g. deblend_psf_instFlux), so we compute the flux
+            # correction factor from any slot flux (it only depends on
+            # position, so any slot with a successful measurement will do)
+            # and apply that to any flux entries that do not have errors.
+            for fluxSlotName in catalog.schema.extract("slot*instFlux"):
+                photoCalibFactor = None
+                for src in catalog:
+                    if np.isfinite(src[fluxSlotName]):
+                        baseSlotName = re.sub("_instFlux", "", fluxSlotName)
+                        photoCalibFactor = (photoCalib.instFluxToNanojansky(src, baseSlotName).value /
+                                            src[fluxSlotName])
+                        break
+                if photoCalibFactor:
+                    catalog[fluxKey] *= photoCalibFactor
```

Comment:
Is flattening the photocalibs for columns without corresponding errors really the best thing we can do?  Would supplying tiny errors work too?